File: Year_2/Semester_4/PROBABILITY_STATISTICS/Examples/clt_simulation.R
# ==============================================================================
# Monte Carlo Simulation of the Central Limit Theorem (CLT) in R
# ==============================================================================
# Demonstrates that the distribution of the sample mean converges to a Normal
# distribution as sample size increases, regardless of the underlying distribution.
# ==============================================================================

set.seed(42)

simulateCentralLimitTheorem <- function(sample_sizes = c(2, 5, 30, 100), num_simulations = 10000) {
  # True parameters for an Exponential distribution with rate lambda = 0.5
  lambda_rate <- 0.5
  true_mean <- 1 / lambda_rate       # E[X] = 2
  true_sd <- 1 / lambda_rate         # SD[X] = 2

  cat("Theoretical Population Mean:", true_mean, "\n")
  cat("Theoretical Population SD:", true_sd, "\n\n")

  results <- list()

  for (n in sample_sizes) {
    # Draws num_simulations samples of size n from Exp(lambda)
    sample_matrix <- matrix(rexp(n * num_simulations, rate = lambda_rate),
                            nrow = num_simulations, ncol = n)

    # Computes sample mean across rows
    sample_means <- rowMeans(sample_matrix)

    # Standardizes the sample means
    theoretical_se <- true_sd / sqrt(n)
    z_scores <- (sample_means - true_mean) / theoretical_se

    results[[as.character(n)]] <- list(
      sample_size = n,
      mean_of_means = mean(sample_means),
      sd_of_means = sd(sample_means),
      expected_se = theoretical_se,
      z_mean = mean(z_scores),
      z_sd = sd(z_scores)
    )

    cat(sprintf("Sample Size n = %3d | Mean: %.4f (Exp: %.4f) | SE: %.4f (Exp: %.4f) | Z-score SD: %.4f\n",
                n, mean(sample_means), true_mean, sd(sample_means), theoretical_se, sd(z_scores)))
  }

  return(results)
}

# Run simulation
cat("Running Central Limit Theorem Monte Carlo Simulation...\n")
sim_results <- simulateCentralLimitTheorem()
cat("Simulation finished.\n")

File: Year_2/Semester_4/PROBABILITY_STATISTICS/Examples/clt_simulation.py
"""Monte Carlo simulation demonstrating the Central Limit Theorem.

Draws repeated random samples from a non-normal exponential distribution,
computes sample means, and illustrates convergence towards a Normal distribution.
"""

import math
import random


def simulateCentralLimitTheorem(
    sample_sizes: list[int] = None, num_simulations: int = 10000
) -> dict:
    """Simulates sample mean convergence under increasing sample sizes.

    Args:
        sample_sizes (list[int]): List of sample sizes n to evaluate.
        num_simulations (int): Number of Monte Carlo iterations per sample size.

    Returns:
        dict: Empirical statistics mapping sample sizes to mean and standard error.
    """
    if sample_sizes is None:
        sample_sizes = [2, 5, 30, 100]

    # Uses an exponential distribution with rate lambda = 0.5.
    lambda_param = 0.5
    true_mean = 1.0 / lambda_param  # Theoretical E[X] = 2.0
    true_sd = 1.0 / lambda_param  # Theoretical SD[X] = 2.0

    print(f"Population Distribution: Exponential(lambda={lambda_param})")
    print(f"Theoretical Mean: {true_mean:.4f}, Theoretical SD: {true_sd:.4f}\n")

    empirical_results = {}
    random.seed(42)

    for n in sample_sizes:
        sample_means = []
        for _ in range(num_simulations):
            # Draws n i.i.d. observations from Exp(lambda).
            sample = [random.expovariate(lambda_param) for _ in range(n)]
            sample_means.append(sum(sample) / n)

        mean_of_means = sum(sample_means) / num_simulations
        variance = sum((x - mean_of_means) ** 2 for x in sample_means) / (
            num_simulations - 1
        )
        sd_of_means = math.sqrt(variance)
        theoretical_se = true_sd / math.sqrt(n)

        empirical_results[n] = {
            "empirical_mean": mean_of_means,
            "empirical_se": sd_of_means,
            "theoretical_se": theoretical_se,
        }

        print(
            f"Sample Size n = {n:3d} | Mean: {mean_of_means:.4f} (True: {true_mean:.4f}) | "
            f"SE: {sd_of_means:.4f} (Theory SE: {theoretical_se:.4f})"
        )

    return empirical_results


if __name__ == "__main__":
    simulateCentralLimitTheorem()

